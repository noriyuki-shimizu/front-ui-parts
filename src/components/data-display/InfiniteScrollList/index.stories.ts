import type { ArgTypeObject, Meta, StoryObj } from '@storybook/vue3'
import { ref } from 'vue'
import InfiniteScrollList from './index.vue'
import type { Props } from './types'

// TODO: Watch: https://github.com/storybookjs/storybook/issues/24238
const InfiniteScrollListComponent = InfiniteScrollList as unknown as Record<string, unknown>

const meta = {
  title: 'Data Display/InfiniteScrollList',
  component: InfiniteScrollListComponent,
  tags: ['autodocs'],
  argTypes: {
    /* slots.emptyDescription */
    emptyDescription: {
      description: '空表示を表すカスタムテキスト',
      control: { type: 'text' },
      table: {
        defaultValue: {
          summary: 'No Data'
        },
        type: { summary: 'string' }
      }
    },
    /* slots.record */
    record: {
      description: 'リスト内のレコード',
      control: 'object',
      table: {
        defaultValue: {
          summary: 'No Data'
        },
        type: { summary: '{ item: T }' }
      }
    }
  },
  args: {
    items: [{ id: '', text: '' }],
    height: '',
    onLoad(page: number) {
      return page === 0
    }
  }
} satisfies Meta<typeof InfiniteScrollListComponent> & ArgTypeObject // TODO: Watch: https://github.com/storybookjs/storybook/issues/23352

export default meta

type Story = StoryObj<typeof meta>

const template: Story = {
  render: (args) => ({
    components: { InfiniteScrollList: InfiniteScrollListComponent },
    setup() {
      const argValue = args as Partial<Props<{ id: string; text: string }[]>>
      const items = ref(argValue.items ?? [])

      const onLoad = async (page: number) => {
        if (items.value.length === 0) {
          return true
        }
        await new Promise((resolve) => {
          setTimeout(() => resolve(null), 2000)
        })
        const newItems = [...Array(10)].map((_, i) => ({
          id: `new-item-${page}-${i}`,
          text: `アイテム ${page * 10 + i}`
        }))
        items.value = [...items.value, ...newItems]
        return items.value.length >= 50
      }

      return { args: argValue, items, onLoad }
    },
    template: `
      <InfiniteScrollList :height="args.height" :items="items" :on-load="onLoad">
        <template #record="{ item }">
          <div style="margin-bottom: 8px;">{{ item.text }}</div>
        </template>
      </InfiniteScrollList>
    `
  })
}

export const Default: Story = {
  ...template,
  args: {
    items: [...Array(10)].map((_, i) => {
      return {
        id: `new-item-0-${i}`,
        text: `アイテム ${i}`
      }
    }),
    height: '300px'
  }
}

export const NoData: Story = {
  ...template,
  args: {
    items: [],
    height: '300px'
  }
}
