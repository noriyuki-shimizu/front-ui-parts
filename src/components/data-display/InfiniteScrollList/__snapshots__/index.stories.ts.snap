// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Display/InfiniteScrollList Default smoke-test 1`] = `
<section tabindex="0"
         class="_infinite-scroll-list_1r7f3_325"
         style="height: 300px;"
>
  <div id="infinite-scroll-list-display-data">
    <div style="margin-bottom: 8px;">
      アイテム 0
    </div>
    <div style="margin-bottom: 8px;">
      アイテム 1
    </div>
    <div style="margin-bottom: 8px;">
      アイテム 2
    </div>
    <div style="margin-bottom: 8px;">
      アイテム 3
    </div>
    <div style="margin-bottom: 8px;">
      アイテム 4
    </div>
    <div style="margin-bottom: 8px;">
      アイテム 5
    </div>
    <div style="margin-bottom: 8px;">
      アイテム 6
    </div>
    <div style="margin-bottom: 8px;">
      アイテム 7
    </div>
    <div style="margin-bottom: 8px;">
      アイテム 8
    </div>
    <div style="margin-bottom: 8px;">
      アイテム 9
    </div>
    <div class="_infinite-scroll-list__spinner-wrapper_1r7f3_331"
         style="display: none;"
    >
      <svg xmlns="http://www.w3.org/2000/svg"
           width="48"
           height="48"
           viewbox="0 0 24 24"
           class="_spinner_aflx1_325 _spinner--default_aflx1_332"
      >
        <path fill-opacity=".1"
              d="M12 0C5.4 0 0 5.4 0 12s5.4 12 12 12 12-5.4 12-12S18.6 0 12 0m0 3c5 0 9 4 9 9s-4 9-9 9-9-4-9-9 4-9 9-9"
        >
        </path>
        <path fill="currentColor"
              d="M12 0c6.6 0 12 5.4 12 12h-3c0-5-4-9-9-9z"
        >
          <animateTransform attributename="transform"
                            dur="1.2s"
                            from="0 12 12"
                            repeatcount="indefinite"
                            to="360 12 12"
                            type="rotate"
          >
          </animateTransform>
        </path>
      </svg>
    </div>
  </div>
</section>
`;

exports[`Data Display/InfiniteScrollList NoData smoke-test 1`] = `
<section tabindex="0"
         class="_infinite-scroll-list_1r7f3_325"
         style="height: 300px;"
>
  <div class="_empty_1ysyh_325 _infinite-scroll-list__empty_1r7f3_328"
       id="infinite-scroll-list-display-empty-data"
  >
    <svg width="120"
         height="120"
         viewbox="0 0 48 48"
    >
      <g fill="currentColor"
         fill-rule="evenodd"
         clip-rule="evenodd"
      >
        <path d="M18.707 4.697a1 1 0 0 1 1.414 0l8.486 8.486a1 1 0 0 1-1.415 1.414L18.707 6.11a1 1 0 0 1 0-1.414">
        </path>
        <path d="M28.707 4.707a1 1 0 0 1 0 1.414l-8.485 8.485a1 1 0 0 1-1.414-1.414l8.485-8.485a1 1 0 0 1 1.414 0M24.316 15.051a1 1 0 0 0-.632 0L8.869 19.99a1 1 0 0 0-.527.41L4.398 25.46a1 1 0 0 0 .473 1.564l3.322 1.107.003 8.631a3 3 0 0 0 2.051 2.845L23.63 44.07a1 1 0 0 0 .747 0l13.382-4.46a3 3 0 0 0 2.051-2.846v-8.63l3.326-1.108a1 1 0 0 0 .472-1.564l-3.995-5.128a1 1 0 0 0-.48-.344zm-14.773 7.06 12.803 4.267-2.698 3.463-12.803-4.267zm14.46 2.71 11.65-3.883L24 17.054l-11.65 3.883zM10.195 36.762l-.003-7.964 9.493 3.163a1 1 0 0 0 1.105-.334l2.21-2.836v12.96l-12.121-4.04a1 1 0 0 1-.684-.95m27.613-7.964-9.49 3.163a1 1 0 0 1-1.105-.334l-2.213-2.84.004 12.963 12.12-4.04a1 1 0 0 0 .684-.948zm-9.451 1.042-2.7-3.464 12.803-4.267 2.7 3.465z">
        </path>
      </g>
    </svg>
    <span class="_empty__description_1ysyh_330">
      No data
    </span>
  </div>
</section>
`;
